tests/components/rest/test_sensor.py

```diff
@@ -1,696 +1,665 @@
+                "headers": {"Accept": "text/xml"},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("sensor.foo")
+    assert state.state == ""
+    assert state.attributes == {"unit_of_measurement": "MB", "friendly_name": "foo"}
+    assert "not a dictionary or list" in caplog.text
+
+
+@respx.mock
+async def test_update_with_json_attrs_bad_JSON(hass, caplog):
+    """Test attributes get extracted from a JSON result."""
+
+    respx.get(
+        "http://localhost",
+        status_code=200,
+        headers={"content-type": CONTENT_TYPE_JSON},
+        content="This is text rather than JSON data.",
+    )
+    assert await async_setup_component(
+        hass,
+        "sensor",
+        {
+            "sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "value_template": "{{ value_json.key }}",
+                "json_attributes": ["key"],
+                "name": "foo",
+                "unit_of_measurement": DATA_MEGABYTES,
+                "verify_ssl": "true",
+                "timeout": 30,
+                "headers": {"Accept": "text/xml"},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("sensor.foo")
+    assert state.state == STATE_UNKNOWN
+    assert state.attributes == {"unit_of_measurement": "MB", "friendly_name": "foo"}
+    assert "Erroneous JSON" in caplog.text
+
+
+@respx.mock
+async def test_update_with_json_attrs_with_json_attrs_path(hass):
+    """Test attributes get extracted from a JSON result with a template for the attributes."""
+
+    respx.get(
+        "http://localhost",
+        status_code=200,
+        headers={"content-type": CONTENT_TYPE_JSON},
+        content='{ "toplevel": {"master_value": "master", "second_level": {"some_json_key": "some_json_value", "some_json_key2": "some_json_value2" } } }',
+    )
+    assert await async_setup_component(
+        hass,
+        "sensor",
+        {
+            "sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "value_template": "{{ value_json.toplevel.master_value }}",
+                "json_attributes_path": "$.toplevel.second_level",
+                "json_attributes": ["some_json_key", "some_json_key2"],
+                "name": "foo",
+                "unit_of_measurement": DATA_MEGABYTES,
+                "verify_ssl": "true",
+                "timeout": 30,
+                "headers": {"Accept": "text/xml"},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+    state = hass.states.get("sensor.foo")
+
+    assert state.state == "master"
+    assert state.attributes["some_json_key"] == "some_json_value"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)